[blog][erlang] I wrote a blog post (#34)
omnicate/omnicate.github.io: pages/blog/_posts/2020-05-18-formatting-erlang.md

---
layout: blogpost
permalink: /blog/formatting-erlang/

> title: Formatting Erlang

Reviewer: Choosing an Erlang formatter?

date: 2020-05-18
tags: erlang rebar3 coding culture
author: <a href="https://www.linkedin.com/in/sebastian-weddmark-olsson/">Sebastian Weddmark Olsson</a>
---

I've looked into different formatters for Erlang and there are some
options. The main two alternatives being `steamroller` and
`rebar3_format`. `steamroller` is slow, which one could have imageined
by its name. `rebar3_format` have issues with macros due to underlying
dependencies.

I've also briefly looked into an Erlang linter called `elvis`.

> 

Reviewer 2: If I understand right, this is the "intro" part? I think a more generic intro will sound better. I notice the other posts we have on the website give a general intro to kinda set the scene. I tried to rephrase a bit the first sentence, see if it makes sense? 
Eg: _There are many different Erlang formatters and I spent some time to look into what are the options that we can use at Working Group Two. The main two alternatives being `steamroller` and
 `rebar3_format`. `steamroller` is slow, which one could have imageined
 by its name. `rebar3_format` have issues with macros due to underlying
 dependencies.

 I've also briefly looked into an Erlang linter called `elvis`._

# Background

In Working Group Two we use a bunch of different programming
languages, and we all have different experiences and are used to
different languages and environments. We are pretty autonomous and we
are expected to jump in and out in different services to fix bugs and
add features.

> We like when the code is uniform because that make it easier to

Reviewer: ```suggestion
We like when the code is uniform, because it makes it easier to
```

> understand it. That is why we want to use tools to make sure our code

Reviewer: ```suggestion
focus on the business logic. That is why we want to use tools to make sure our code
```

> is consistent no matter who writes it, which tool, or regards to

Reviewer: which tools they use* ? I don't understand the last part of the sentence ("or regards to language rules or background.")

language rules or background.

About half a year ago there was a discussion about code style within

> WG2 that resulted in formatting tools being applied for Kotlin, Bazel,

Reviewer: We agreed to always use `Working Group Two` (the full name) when referring to ourselves.

Go and Java. It also resulted in an internal wiki page containing
guidelines about code style.

> That document highlights some of the problems with different

Reviewer: "That document highlights some of the problems with different
code-styles mixed."

Should this be

"That document highlights some of the problems with mixing different code-styles" ?

The requirements I had was that it should be reproducable. Calling the
formatter multiple times should not change the structure more than
once (first time of being called). The formatter should also
preferably work with rebar3 (the most used Erlang build tool).  The
tool should not use external tooling that wouldn't work for all
developers flow.

I also wanted it to have a short execution time, at least after the
initial formatting.

# Benchmark

I searched for formatting tools on hex.pm, github.com, duckduckgo.com,
google.com and came up with the following arbitrary list of Erlang
formatters. There is probably others, but these seems to be the most
used.

- rebar3_fmt
- steamroller
- otp/erl_tidy
- tsloughter/erl_tidy
- rebar3_format
- eryngii

## [rebar3_fmt](https://github.com/fenollp/erlang-formatter)

One big problem with this is that it uses Emacs `erlang-mode` for
formatting. Sure, I am an Emacs user and the `erlang-mode` and its
formatting is maintained and supperted by OTP, but my non-emacs
coworkers would not be happy if they need to install Emacs every time
they want to format the code.

## [steamroller](https://github.com/old-reliable/steamroller/)

> Though it was easy to setup, just add it to dependencies in your Rebar
> config and run `rebar3 steamroll`, my first impression of the

Reviewer: ```suggestion
Though it was easy to setup (just add it to dependencies in your Rebar
config and run `rebar3 steamroll`), my first impression of the
```

execution was that it was really slow. Even when running subsequent

> calls on my Dell XPS 13 P82G it took around 3.5 minutes to format.

Reviewer: > it took around 3.5 minutes to format.

Ouch!

The plugin had some support for increasing the number of workers from
the default `--J=1`, but that did not seem to help with the execution
time.

The default steamroller formatting options specify 2 spaces instead of
the `erlang-mode` 4 spaces standard that is used in our code base.

Here is a sample of a complex record structure

```

> -                                  components =

Reviewer: Is this indentation correct?
